File: code/lua/exercises_test.lua
require "exercises"

--------------------------------------------------------------------------------
passed, failed = 0, 0

function suite(name)
  io.write("\nTesting " .. name)
end

function expect(condition)
  if condition then passed = passed + 1 else failed = failed + 1 end
  io.write(condition and " ." or " F")
end

function expect_array_equals(a1, a2)
  if #a1 ~= #a2 then expect(false); return end
  for i = 1, #a1 do
    if a1[i] ~= a2[i] then expect(false); return end
  end
  expect(true)
end

function expect_hash_equals(h1, h2)
  for k, v in pairs(h1) do
    if h2[k] ~= v then expect(false); return end
  end
  for k, v in pairs(h2) do
    if h1[k] ~= v then expect(false); return end
  end
  expect(true)
end

function expect_error(text, fun, arg)
  local success, err = pcall(fun, arg)
  expect(not success and string.match(err, text) ~= nil)
end
--------------------------------------------------------------------------------

suite("first_then_lower_case")
function nonEmpty(s) return s ~= "" end
function lengthGreaterThan3(s) return #s > 3 end
function to_lower(s) return string.lower(s) end
function starts_with_b(s) return string.lower(string.sub(s, 1, 1)) == "b" end

expect(first_then_apply({}, nonEmpty, to_lower) == nil)
expect(first_then_apply({"", "A", "B"}, nonEmpty, to_lower) == "a")
expect(first_then_apply({"", "A", "ABC"}, lengthGreaterThan3, to_lower) == nil)
expect(first_then_apply({"ABC", "ABCD", "ABCDE"}, lengthGreaterThan3, to_lower) == "abcd")
expect(first_then_apply(
    {"ABC", "ABCD", "ABCDE", "Bee"}, starts_with_b, to_lower) == "bee")
expect(first_then_apply(
    {"ABC", "ABCD", "ABCDE", "Bee"}, starts_with_b, function(s) return #s end) == 3)

suite("powers_generator")
gen = powers_generator(3, 100)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == 1)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == 3)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == 9)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == 27)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == 81)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "suspended" and success and power == nil)
status = coroutine.status(gen)
success, power = coroutine.resume(gen)
expect(status == "dead" and not success)

suite("say")
expect(say() == "")
expect(say("hi")() == "hi")
expect(say("hi")("there")() == "hi there")
expect(say("hello")("my")("name")("is")("Colette")() == "hello my name is Colette")
expect(say("h i")() == "h i")
expect(say("hi ")("   there")() == "hi     there")
expect(say("")("")("dog")("")("go")() == "  dog  go")
expect(say("😄🤗")("💀👊🏾")() == "😄🤗 💀👊🏾")

suite("meaningful_line_count")
expect_error("No such file", meaningful_line_count, "no-such-file.txt")
expect(meaningful_line_count("../../test-data/test-for-line-count.txt") == 5)

suite("Quaternion")
q = Quaternion.new(3.5, 2.25, -100.0, -1.25)
expect(q.a == 3.5)
expect(q.b == 2.25)
expect(q.c == -100.0)
expect(q.d == -1.25)
q1 = Quaternion.new(1.0, 3.0, 5.0, 2.0);
q2 = Quaternion.new(-2.0, 2.0, 8.0, -1.0);
q3 = Quaternion.new(-1.0, 5.0, 13.0, 1.0);
q4 = Quaternion.new(-46.0, -25.0, 5.0, 9.0);
zero = Quaternion.new(0, 0, 0, 0)
i = Quaternion.new(0, 1, 0, 0)
j = Quaternion.new(0, 0, 1, 0)
k = Quaternion.new(0, 0, 0, 1)
expect_array_equals(zero:coefficients(), {0.0, 0.0, 0.0, 0.0})
expect_array_equals(k:coefficients(), {0.0, 0.0, 0.0, 1.0})
expect_array_equals(q2:coefficients(), {-2.0, 2.0, 8.0, -1.0})
expect(q1 + q2 == q3)
expect(q1 * q2 == q4)
expect(q1 + zero == q1)
expect(q1 * zero == zero)
expect(i * j == k)
expect(j * k == i)
expect(j + i == Quaternion.new(0.0, 1.0, 1.0, 0.0))
expect(tostring(zero) == "0")
expect(tostring(j) == "j")
expect(tostring(k:conjugate()) == "-k")
expect(tostring(j:conjugate() * Quaternion.new(2.0, 0.0, 0.0, 0.0)) == "-2.0j")
expect(tostring(j + k) == "j+k")
expect(tostring(Quaternion.new(0.0, -1.0, 0.0, 2.25)) == "-i+2.25k")
expect(tostring(Quaternion.new(-20.0, -1.75, 13.0, -2.25)) == "-20.0-1.75i+13.0j-2.25k")
expect(tostring(Quaternion.new(-1.0, -2.0, 0.0, 0.0)) == "-1.0-2.0i")
expect(tostring(Quaternion.new(1.0, 0.0, -2.0, 5.0)) == "1.0-2.0j+5.0k")

print(string.format("\n%d passed, %d failed", passed, failed))

File: code/java/ExercisesTest.java
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.io.IOException;
import java.io.FileNotFoundException;
import java.util.concurrent.Callable;

public class ExercisesTest {

    //////////////////////////////////////////////////////////////////////////////////////
    HashMap<Boolean, Integer> counts = new HashMap<>(Map.of(true, 0, false, 0));

    void suite(String name) {
        IO.print("\nTesting " + name);
    }

    void expect(boolean condition) {
        counts.put(condition, counts.get(condition) + 1);
        IO.print(condition ? " ." : " F");
    }

    void expectToThrow(Callable<?> c, Class<? extends Exception> exception, String message) {
        try {
            c.call();
            expect(false);
        } catch (Exception e) {
            expect(exception.isInstance(e) && e.getMessage().contains(message));
        }
    }
    //////////////////////////////////////////////////////////////////////////////////////

    void main() throws IOException {

        suite("firstThenLowerCase");
        expect(Exercises.firstThenLowerCase(
                List.of(), s -> !s.isEmpty()).isEmpty());
        expect(Exercises.firstThenLowerCase(
                List.of("", "A", "B"), s -> !s.isEmpty()).get().equals("a"));
        expect(Exercises.firstThenLowerCase(
                List.of("", "A", "ABC"), s -> s.length() > 3).isEmpty());
        expect(Exercises.firstThenLowerCase(
                List.of("ABC", "ABCD", "ABCDE"), s -> s.length() > 3).get().equals("abcd"));

        suite("say");
        expect(Exercises.say().phrase().equals(""));
        expect(Exercises.say("hi").phrase().equals("hi"));
        expect(Exercises.say("Oh").and("kay").phrase().equals("Oh kay"));
        expect(Exercises.say("hello").and("my").and("name").and("is").and("Colette").phrase()
                .equals("hello my name is Colette"));
        expect(Exercises.say("h i").phrase().equals("h i"));
        expect(Exercises.say("hi ").and("   there").phrase().equals("hi     there"));
        expect(Exercises.say("").and("").and("dog").and("").and("go").phrase()
                .equals("  dog  go"));
        expect(Exercises.say("🐤🦇").and("$🦊👏🏽").and("!").phrase().equals("🐤🦇 $🦊👏🏽 !"));
        expect(Exercises.say("😄🤗").and("💀👊🏾").phrase().equals("😄🤗 💀👊🏾"));
        // Ensure there is no sharing of partial states!
        var greet = Exercises.say("Hello").and("there");
        expect(greet.and("nice").and("person").phrase().equals("Hello there nice person"));
        expect(greet.and("Swift").phrase().equals("Hello there Swift"));

        suite("meaningfulLineCount");
        expectToThrow(
            () -> Exercises.meaningfulLineCount("no-such-file.txt"),
            FileNotFoundException.class,
            "No such file");
        expect(Exercises.meaningfulLineCount("../../test-data/test-for-line-count.txt") == 5);

        suite("Quaternion");
        expectToThrow(() -> new Quaternion(Double.NaN, 0, 0, 0),
                IllegalArgumentException.class, "Coefficients cannot be NaN");
        expectToThrow(() -> new Quaternion(0, Double.NaN, 0, 0),
                IllegalArgumentException.class, "Coefficients cannot be NaN");
        expectToThrow(() -> new Quaternion(0, 0, Double.NaN, 0),
                IllegalArgumentException.class, "Coefficients cannot be NaN");
        expectToThrow(() -> new Quaternion(0, 0, 0, Double.NaN),
                IllegalArgumentException.class, "Coefficients cannot be NaN");

        var q = new Quaternion(3.5, 2.25, -100, -1.25);
        expect(q.a() == 3.5);
        expect(q.b() == 2.25);
        expect(q.c() == -100.0);
        expect(q.d() == -1.25);

        var q1 = new Quaternion(1, 3, 5, 2);
        var q2 = new Quaternion(-2, 2, 8, -1);
        var q3 = new Quaternion(-1, 5, 13, 1);
        var q4 = new Quaternion(-46, -25, 5, 9);

        expect(Quaternion.ZERO.coefficients().equals(List.of(0.0, 0.0, 0.0, 0.0)));
        expect(Quaternion.K.coefficients().equals(List.of(0.0, 0.0, 0.0, 1.0)));
        expect(q2.coefficients().equals(List.of(-2.0, 2.0, 8.0, -1.0)));

        expect(q1.plus(q2).equals(q3));
        expect(q1.times(q2).equals(q4));
        expect(q1.plus(Quaternion.ZERO).equals(q1));
        expect(q1.times(Quaternion.ZERO).equals(Quaternion.ZERO));

        expect(Quaternion.I.times(Quaternion.J).equals(Quaternion.K));
        expect(Quaternion.J.times(Quaternion.K).equals(Quaternion.I));
        expect(Quaternion.J.plus(Quaternion.I).equals(new Quaternion(0, 1, 1, 0)));

        expect(("" + Quaternion.ZERO).equals("0"));
        expect(("" + Quaternion.J).equals("j"));
        expect(("" + Quaternion.K.conjugate()).equals("-k"));
        expect(("" + Quaternion.J.conjugate().times(new Quaternion(2, 0, 0, 0))).equals("-2.0j"));
        expect(("" + Quaternion.J.plus(Quaternion.K)).equals("j+k"));
        expect(("" + new Quaternion(0, -1, 0, 2.25)).equals("-i+2.25k"));
        expect(("" + new Quaternion(-20, -1.75, 13, -2.25)).equals("-20.0-1.75i+13.0j-2.25k"));
        expect(("" + new Quaternion(-1, -2, 0, 0)).equals("-1.0-2.0i"));
        expect(("" + new Quaternion(1, 0, -2, 5)).equals("1.0-2.0j+5.0k"));

        suite("BinarySearchTree");
        BinarySearchTree t = new Empty();
        expect(t.size() == 0);
        expect(!t.contains("A"));
        expect(t.toString().equals("()"));
        t = t.insert("G");
        expect(t.size() == 1);
        expect(t.contains("G"));
        expect(!t.contains("A"));
        expect(t.toString().equals("(G)"));
        t = t.insert("B");
        expect(t.toString().equals("((B)G)"));
        t = t.insert("D");
        expect(t.toString().equals("((B(D))G)"));
        t = t.insert("H");
        expect(t.toString().equals("((B(D))G(H))"));
        t = t.insert("A");
        expect(t.toString().equals("(((A)B(D))G(H))"));
        t = t.insert("C");
        t = t.insert("J");
        expect(t.size() == 7);
        expect(t.contains("J"));
        expect(!t.contains("Z"));
        expect(t.toString().equals("(((A)B((C)D))G(H(J)))"));
        // Test immutability
        BinarySearchTree t2 = t;
        t2 = t2.insert("F");
        expect(t2.size() == 8);
        expect(t.size() == 7);

        IO.println(String.format(
                "\n%d passed, %d failed",
                counts.get(true),
                counts.get(false)
        ));
    }
}
